shared/mad/codeql/mad/modelgenerator/ModelGeneratorImpl.qll

```diff
@@ -0,0 +1,909 @@
+  }
+
+  final private class SummaryTargetApiFinal = SummaryTargetApi;
+
+  class DataFlowSummaryTargetApi extends SummaryTargetApiFinal {
+    DataFlowSummaryTargetApi() { not isUninterestingForDataFlowModels(this) }
+  }
+
+  class DataFlowSourceTargetApi = SourceTargetApi;
+
+  class DataFlowSinkTargetApi = SinkTargetApi;
+
+  private module ModelPrintingInput implements Printing::ModelPrintingSig {
+    class SummaryApi = DataFlowSummaryTargetApi;
+
+    class SourceOrSinkApi = SourceOrSinkTargetApi;
+
+    string getProvenance() { result = "df-generated" }
+  }
+
+  module ModelPrinting = Printing::ModelPrinting<ModelPrintingInput>;
+
+  /**
+   * Holds if `c` is a relevant content kind, where the underlying type is relevant.
+   */
+  private predicate isRelevantTypeInContent(DataFlow::ContentSet c) {
+    isRelevantType(getUnderlyingContentType(c))
+  }
+
+  /**
+   * Holds if data can flow from `node1` to `node2` either via a read or a write of an intermediate field `f`.
+   */
+  private predicate isRelevantTaintStep(DataFlow::Node node1, DataFlow::Node node2) {
+    exists(DataFlow::ContentSet f |
+      DataFlow::readStep(node1, f, node2) and
+      // Partially restrict the content types used for intermediate steps.
+      (not exists(getUnderlyingContentType(f)) or isRelevantTypeInContent(f))
+    )
+    or
+    exists(DataFlow::ContentSet f | DataFlow::storeStep(node1, f, node2) | containerContent(f))
+  }
+
+  /**
+   * Holds if content `c` is either a field, a synthetic field or language specific
+   * content of a relevant type or a container like content.
+   */
+  pragma[nomagic]
+  private predicate isRelevantContent0(DataFlow::ContentSet c) {
+    isRelevantTypeInContent(c) or
+    containerContent(c)
+  }
+
+  /**
+   * Gets the MaD string representation of the parameter node `p`.
+   */
+  string parameterNodeAsInput(DataFlow::ParameterNode p) {
+    result = parameterAccess(p.(NodeExtended).asParameter())
+    or
+    result = qualifierString() and p instanceof InstanceParameterNode
+  }
+
+  /**
+   * Gets the MaD string representation of the parameter `p`
+   * when used in content flow.
+   */
+  string parameterNodeAsContentInput(DataFlow::ParameterNode p) {
+    result = parameterContentAccess(p.(NodeExtended).asParameter())
+    or
+    result = qualifierString() and p instanceof InstanceParameterNode
+  }
+
+  /**
+   * Gets the MaD input string representation of `source`.
+   */
+  string asInputArgument(NodeExtended source) { result = getInputArgument(source) }
+
+  /**
+   * Gets the summary model of `api`, if it follows the `fluent` programming pattern (returns `this`).
+   */
+  string captureQualifierFlow(DataFlowSummaryTargetApi api) {
+    exists(ReturnNodeExt ret |
+      api = returnNodeEnclosingCallable(ret) and
+      isOwnInstanceAccessNode(ret)
+    ) and
+    result = ModelPrinting::asLiftedValueModel(api, qualifierString(), "ReturnValue")
+  }
+
+  private int accessPathLimit0() { result = 2 }
```

Comment:
I think it would be nice to put all logic related to content-sensitive model generation inside a nested module, say `ContentSensitive`.